File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        prefix_sum = 0
        prefix_sum_count = {0: 1}  
        
        for num in nums:
            prefix_sum += num 
            if (prefix_sum - k) in prefix_sum_count:
                count += prefix_sum_count[prefix_sum - k]  
            if prefix_sum in prefix_sum_count:
                prefix_sum_count[prefix_sum] += 1  
            else:
                prefix_sum_count[prefix_sum] = 1 
        
        return count
        
        # result=0
        # for i in range(len(nums)):
        #     summ=0
        #     for j in range(i,len(nums)):
        #         summ=summ+nums[j]
        #         if summ==k:
        #             result=result+1
        #             break
        #         elif summ>k:
        #             break    
        # return result             


        # count=0
        # for i in range(len(nums)):
        #     c=nums[i]
        #     if c==k:
        #         count=count+1
        #     for j in range(i+1,len(nums)):
        #         c=c+nums[j]
        #         if c==k:
        #             count=count+1
        # return count       

        

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int result=0;
        Map<Integer,Integer> hm=new HashMap<>();
        int sum=0;
        for(int num:nums)
        {
            sum=sum+num;
            if(sum==k){
                result=result+1;
            }
            if(hm.containsKey(sum-k)){
                result=result+hm.get(sum-k);
            }
            if(hm.containsKey(sum)){
                hm.put(sum,hm.get(sum)+1);
            }else{
                hm.put(sum,1);
            }
        }return result;



    //     int result=0;
    //     for(int i=0;i<nums.length;i++)
    //     {
    //         int sum=0;
    //         for(int j=i;j<nums.length;j++)
    //         {
    //             sum=sum+nums[j];
    //             if(sum==k){
    //                 result++;
    //             }
    //         }
    //     }return result;
        
    }
}